### PR TITLE
Oris 2867 scw auth script

### PR DESCRIPTION
**Pull request by oren-levin**

### CODE REVIEW

File: scripts/securechange-authorize-change.sh [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+  echo "  TUFIN_SECURECLOUD_API_KEY      - API key to Tufin SecureCloud with global admin permissions"
+  echo "  TUFIN_SECURECHANGE_URL         - Tufin SecureChange URL"
+  echo "  TUFIN_SECURECHANGE_CREDENTIALS - Tufin SecureChange credentials in the format of username:password"
+  echo ""
+  echo "Mandatory arguments"
+  echo "  ticketId           - SecureChange ticket ID to authorize"
+  echo "  terraformPlanFile  - Path to terraform plan file"
+  echo "  scwResponseFile    - Path to temp file that will hold the response from SecureChangeWorkflow"
+  echo "  scTargetsFile      - Path to temp file that will hold the possible target IDs from SecureCloud"
+  echo "  scAuthResponseFile - Path to temp file that will hold the authorization response from SecureCloud"
```
Comment (niratzitz) [inline scripts/securechange-authorize-change.sh:18]: do we really need the user to provide all these file names ? (scAuthRequestBody is missing from the list...)
maybe for simplicity we can initialize all these variables to hard-coded file names ?

File: scripts/securechange-authorize-change.sh [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+ExitWithUsage() {
+  PrintUsage
+  exit 1
+}
+
+GetTicketFromScw() {
+  getTicketUrl="${TUFIN_SECURECHANGE_URL}/api/securechange/tickets/${ticketID}.json"
+  echo "Calling api: ${getTicketUrl}"
```
Comment (niratzitz) [inline scripts/securechange-authorize-change.sh:98]: consider adding the action to these printouts like "Getting ticket $ticketID from SecureChange. Calling API: $getTicketUrl"

File: scripts/securechange-authorize-change.sh [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+  fi
+
+  targetsJqSelector='.ticket.steps.step[] | select(.name == "Technical Design") | .tasks.task.fields.field[].access_request.targets.target[]? | select(.external_uid) | .external_uid'
+  scwTargets=$(jq "${targetsJqSelector}" "${scwResponseFile}" | paste -s -d',')
+
+  sourcesJqSelector='.ticket.steps.step[] | select(.name == "Technical Design") | .tasks.task.fields.field[].access_request.sources.source[]? | [.ip_address, .cidr|tostring] | join("/")'
+  readarray -t scwSources <<<$(jq -r "${sourcesJqSelector}" "${scwResponseFile}")
+
+  destinationsJqSelector='.ticket.steps.step[] | select(.name == "Technical Design") | .tasks.task.fields.field[].access_request.destinations.destination[]? | [.ip_address, .cidr|tostring] | join("/")'
+  readarray -t scwDestinations <<<$(jq -r "${destinationsJqSelector}" "${scwResponseFile}")
```
Comment (niratzitz) [inline scripts/securechange-authorize-change.sh:120]: add `readarray` to the `tools` list of `ValidateDependencies`

File: scripts/securechange-authorize-change.sh [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+  anyServiceCount=$(jq -r "${isAnyJqSelector}" "${scwResponseFile}" | wc -l)
+  if [[ anyServiceCount -eq 0 ]]; then
+    servicesJqSelector='.ticket.steps.step[] | select(.name == "Technical Design") | .tasks.task.fields.field[].access_request.services.service[]? | [.protocol, .port|tostring] | @csv'
+    readarray -t scwServicesCsv <<<$(jq -r "${servicesJqSelector}" "${scwResponseFile}")
+  fi
+}
+
+FindTargetIds() {
+  getTargetsUrl="${TUFIN_SECURECLOUD_URL}/iris/api/provisioning/targets"
+  echo "Calling api: ${getTargetsUrl}"
```
Comment (niratzitz) [inline scripts/securechange-authorize-change.sh:132]: consider adding the action to these printouts like "Getting targets from SecureCloud. Calling API: $getTicketUrl"

File: scripts/securechange-authorize-change.sh [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+  code=$(curl -s -k -w "%{response_code}" -H "Authorization: Bearer ${TUFIN_SECURECLOUD_API_KEY}" -o "${scTargetsFile}" "${getTargetsUrl}")
+  if [[ "${code}" -ne "200" ]]; then
+    echo "ERROR: SecureChange HTTP response status code was ${code}, URL: ${getTicketUrl}"
```
Comment (niratzitz) [inline scripts/securechange-authorize-change.sh:135]: typo - it should be SecureCloud HTTP response here...

File: scripts/securechange-authorize-change.sh [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+  fi
+
+  targetIDsJqSelector=".targets[] | select(.azure_vnet_id==(${scwTargets})) | .id"
+  readarray -t scTargetIDs <<<$(jq -r "${targetIDsJqSelector}" "${scTargetsFile}")
+  if [[ ${#scTargetIDs[@]} -eq 0 ]]; then
+    echo "ERROR: Did not find any target"
+    exit 1
+  else
+    echo "Found ${#scTargetIDs[@]} targets"
+  fi
```
Comment (niratzitz) [inline scripts/securechange-authorize-change.sh:146]: consider adding "SecureCloud" to this printouts so it is more clear

File: scripts/securechange-authorize-change.sh [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+    done
+  else
+    authRequestBody+='{"type":"ANY_NETWORK"}'
+  fi
+  authRequestBody+=']'
+}
+
+PostAuthRequest() {
+  postAuthUrl="${TUFIN_SECURECLOUD_URL}/api/iris/model/cross-account/integration/verify-changes"
+  echo "Calling api: ${postAuthUrl}"
```
Comment (niratzitz) [inline scripts/securechange-authorize-change.sh:212]: consider adding the action to these printouts like "Verifying terraform plan is authorized by the access-request of ticket $ticketID using SecureCloud API: $getTicketUrl"

File: scripts/securechange-authorize-change.sh [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+  if [[ -z "${scAuthRequestBody}" ]]; then
+    echo "ERROR: Please provide a tmp file to save SC authorization request body"
+    ExitWithUsage
+  fi
+
+  if [[ -z "${scAuthResponseFile}" ]]; then
+    echo "ERROR: Please provide a tmp file to save SC authorization response"
+    ExitWithUsage
+  fi
```
Comment (niratzitz) [inline scripts/securechange-authorize-change.sh:88]: if we initialize all these variables to hard-coded values we can omit these checks

**Review by niratzitz — CHANGES_REQUESTED** [review]

File: scripts/securechange-authorize-change.sh (hunk 1 above)
Comment (oren-levin) [reply scripts/securechange-authorize-change.sh:18]: Well, the "user" is us... and I think it's better to keep the script unaware to the environment and control everything from the pipeline definitions so we can keep/erase/debug whatever we need without changing the code.
I'll fix the help for scAuthRequestBody .

File: scripts/securechange-authorize-change.sh (hunk 2 above)
Comment (oren-levin) [reply scripts/securechange-authorize-change.sh:98]: Fixed

File: scripts/securechange-authorize-change.sh (hunk 3 above)
Comment (oren-levin) [reply scripts/securechange-authorize-change.sh:120]: AFAIK it's a bash command

**Review by niratzitz — APPROVED** [review]

**Pull request merged**